In [ ]:
from google.colab import drive
import pandas as pd
import numpy as np
from google.colab import drive
drive.mount('/content/drive')
file_path='/content/drive/My Drive/datasets/unsupervised_data (2).csv'
data=pd.read_csv(file_path)
matrix = data.iloc[:, 1:].values

Mounted at /content/drive


In [ ]:
import numpy as np
import matplotlib.pyplot as plt

X = np.array(matrix)
for r in range(X.shape[1]):
  X[:, r] = (X[:, r] - np.mean(X[:, r])) / np.std(X[:, r])

def kmeans(X, k, max_iter=100, seed=1):
  if seed+1:
    np.random.seed(seed)
  centroids = X[np.random.choice(X.shape[0], k, replace=False)]

  for _ in range(max_iter):
    distances = np.sqrt(((X[:, np.newaxis] - centroids) ** 2).sum(axis=2))

    labels = np.argmin(distances, axis=1)


    new_centroids = np.array([
      X[labels == i].mean(axis=0) if np.any(labels == i) else centroids[i]
      for i in range(k)
    ])

    if np.allclose(centroids, new_centroids, atol=1e-4):
      break

    centroids = new_centroids


  wcss = sum(np.sum((X[labels == i] - centroids[i]) ** 2) for i in range(k))
  return centroids, labels, wcss

def multifold_kmeans(X, k, n_folds=10, max_iter=100):
  best_centroids = None
  best_labels = None
  best_wcss = float('inf')

  for fold in range(n_folds):
    centroids, labels, wcss = kmeans(X, k, max_iter, seed=fold)
    if wcss < best_wcss:
      best_wcss = wcss
      best_labels = labels
      best_centroids = centroids

  return best_centroids, best_labels, best_wcss


for k in range(1, 16):
  best_centroids, best_labels, best_wcss = multifold_kmeans(X, k, n_folds=10, max_iter=100)
  print(f"Best WCSS for k={k}: {best_wcss}")


Best WCSS for k=1: 240000.0
Best WCSS for k=2: 154874.15114342183
Best WCSS for k=3: 105277.72120411889
Best WCSS for k=4: 61805.662473229786
Best WCSS for k=5: 32651.439918803644
Best WCSS for k=6: 13389.876039813465
Best WCSS for k=7: 6388.010965497643
Best WCSS for k=8: 6241.985361776473
Best WCSS for k=9: 6099.82274795597
Best WCSS for k=10: 5952.72138811445
Best WCSS for k=11: 5809.495535815292
Best WCSS for k=12: 5668.593539777473
Best WCSS for k=13: 5663.852271943992
Best WCSS for k=14: 5519.309590673804
Best WCSS for k=15: 5373.385996087657


Best WCSS for k=2: 154874.15114342183
Best WCSS for k=3: 105277.7212041189
Best WCSS for k=4: 61805.66247322979
Best WCSS for k=5: 32651.439918803648
Best WCSS for k=6: 13389.876039813467
Best WCSS for k=7: 6388.010965497643
Best WCSS for k=8: 6238.435872266065
Best WCSS for k=9: 6092.410574432229
Best WCSS for k=10: 5949.184722133069



After k=7, the value of wcss decreases very slowly. hence, k=7 is the ideal value of k, with wcss= 6388. However, even after multiple initialisation, with other hyperparameters same, the elbow value of k varies between 7-9